### ISSUE TITLE
[BUG] Broken links in "Extra" menu

### ISSUE THREAD

**Issue by artggd** [bolt/bolt#6139]
The links in the _Extra_ menu are malformed since the last update. They don't work when using custom parameters on every route (example: _\_locale_). The URLs are generated like `/admin?_locale=nl/extend`

Details
-------

 - Relevant Bolt Version: dev-release/3.3
 - Install type:  Composer install
 - PHP version: 7.0

Reproduction
------------

 - Install [bolt-translate](https://github.com/AnimalDesign/bolt-translate/issues)
 - Configure it with a few languages
 - Click on one of the links in _Extra_ menu

I found the source of the bug, will PR asap.

**Comment by bobdenotter** [comment]
> I found the source of the bug, will PR asap.

As before: Excellent, looking forward to it! 👍

**Comment by artggd** [comment]
Right. So, I dug quite a bit and I am still unsure where this should be fixed.

The bug appears for the following reasons:
1. The links in the _extend_ menu are built by [concatenating pieces of urls](https://github.com/bolt/bolt/blob/master/src/Menu/MenuEntry.php#L64)
1. The root url (/admin) used to be a string but [is now a generated route](https://github.com/bolt/bolt/commit/2380a4515089e2465bc30a60043eda0d688412f6#diff-87d8ffd8f6b681ed884dccac349951e6L31)
1. bolt-translate extension wraps _url_generator_ service to add a _\_locale_ parameter in the querystring, each time a route is generated

The bug could be fixed on any level of that stack. 

It's definitely way easier to fix it on bolt-translate's side, but I reckon concatenating pieces of urls is dodgy, knowing that those urls are generated by _url_generator_ and could contain a query string. I understand this is also part of bolt's architecture though.

What do you think? Where should this be fixed?

**Comment by bobdenotter** [comment]
Pinging @CarsonF .. Opinions on this?

**Comment by CarsonF** [comment]
Hmmm. I would probably say we should assume urls generated from `url_generator` are final and not try to concat them. 

I was trying to get rid of `ResourceManager::getUrl` so using that is out of the question.
Two options I see to change the menu builder:
1. Concat the request base path and the branding var as the base path. 
2. Define a route we can use to generate a final url.

I think the second option is better as the first doesn't ever call the `url_generator` so any extra query params added there would not be applied.

**Comment by artggd** [comment]
Why not both? 🌵 

First option would be great as a quick and easy fix to help stabilise this version.
Second option could be kept for the next version as it would require some refactoring and may introduce BC breaks (I guess?).

I thought of a third solution: stripping-out the query string before the concat then rebuild it on the fly and eventually put it at the end of the url. I'm not so keen on this one cause it's quite hacky but it would be an intermediate solution.

**Comment by CarsonF** [comment]
The first option would also break the links as well, since the `_locale` query parameter would never get added to the url. 

The second option isn't that hard. I want to talk about the implementation with @GawainLynch though since he wrote it.

But basically it would be a route: 
```php
$ctr->get('{path}')
->assert('path', '.+')
->bind('admin_menu');
```

Then `MenuEntry::getUri` would change to generate that route. The path would be like a method like the current `getUri`, just without the base path applied to the parent. 

The only problem I see is we don't want to match that route when doing request matching.

**Comment by GawainLynch** [comment]
Fixed in #6142
